pkg/storage/azurecdn/storage.go

```diff
@@ -49,49 +52,88 @@ func (s *Storage) Save(ctx context.Context, module, version string, mod []byte,
 	// This container must exist
 	containerURL := serviceURL.NewContainerURL("gomodules")
 
-	infoBlobURL := containerURL.NewBlockBlobURL(fmt.Sprintf("%s/@v/%s.info", module, version))
-	modBlobURL := containerURL.NewBlockBlobURL(fmt.Sprintf("%s/@v/%s.mod", module, version))
-	zipBlobURL := containerURL.NewBlockBlobURL(fmt.Sprintf("%s/@v/%s.zip", module, version))
-
+	emptyMeta := map[string]string{}
+	emptyBlobAccessCond := azblob.BlobAccessConditions{}
 	httpHeaders := func(contentType string) azblob.BlobHTTPHeaders {
 		return azblob.BlobHTTPHeaders{
 			ContentType: contentType,
 		}
 	}
-	emptyMeta := map[string]string{}
-	emptyBlobAccessCond := azblob.BlobAccessConditions{}
+	uploadOpts := func(contentType string) uploadOptions {
+		return uploadOptions{
+			accessConditions: emptyBlobAccessCond,
+			metadata:         emptyMeta,
+			blobHTTPHeaders:  httpHeaders(contentType),
+		}
+	}
 
 	const numUpload = 3
-	uploadErrs := make(chan error, numUpload)
+	errChan := make(chan error, numUpload)
 
-	upload := func(url azblob.BlockBlobURL, content io.ReadSeeker, contentType string) {
-		_, err := url.Upload(ctx, content, httpHeaders(contentType), emptyMeta, emptyBlobAccessCond)
-		uploadErrs <- err
-	}
-	zipBytes, err := ioutil.ReadAll(zip)
-	if err != nil {
-		return err
-	}
-
-	go upload(infoBlobURL, bytes.NewReader(info), "application/json")
-	go upload(modBlobURL, bytes.NewReader(mod), "text/plain")
-	go upload(zipBlobURL, bytes.NewReader(zipBytes), "application/octet-stream")
+	go uploadBlob(tctx, errChan, containerURL, module, version, "info", bytes.NewReader(info), uploadOpts("application/json"))
+	go uploadBlob(tctx, errChan, containerURL, module, version, "mod", bytes.NewReader(mod), uploadOpts("text/plain"))
+	go uploadStream(tctx, errChan, containerURL, module, version, "zip", zip, uploadOpts("application/octet-stream"))
 
 	var errors error
 	for i := 0; i < numUpload; i++ {
-		select {
-		case err := <-uploadErrs:
-			if err != nil {
-				errors = multierror.Append(errors, err)
-			}
-		case <-ctx.Done():
-			return ctx.Err()
+		err := <-errChan
+		if err != nil {
+			errors = multierror.Append(errors, err)
 		}
 	}
+	close(errChan)
 
 	// TODO: take out lease on the /list file and add the version to it
 	//
 	// Do that only after module source+metadata is uploaded
 
 	return errors
 }
+
+func blockBlobURL(containerURL azblob.ContainerURL, module, version, name string) azblob.BlockBlobURL {
+	return containerURL.NewBlockBlobURL(fmt.Sprintf("%s/@v/%s.%s", module, version, name))
```

Comment:
Here you can use `config.PackageVersionedName(module, version, name)` instead of `fmt.Sprintf`